File: content/manual/v8.1/source/includes/aggregation/aggregation-examples/group-and-total/full-files/group_total.rb
# typed: strict
require "mongo"
require "bson"

uri = "<connection string>"

Mongo::Client.new(uri) do |client|
  agg_db = client.use("agg_tutorials_db")

  # start-insert-orders
  orders = agg_db[:orders]
  orders.delete_many({})

  orders.insert_many(
    [
      {
        customer_id: "elise_smith@myemail.com",
        orderdate: DateTime.parse("2020-05-30T08:35:52Z"),
        value: 231,

      },
      {
        customer_id: "elise_smith@myemail.com",
        orderdate: DateTime.parse("2020-01-13T09:32:07Z"),
        value: 99,
      },
      {
        customer_id: "oranieri@warmmail.com",
        orderdate: DateTime.parse("2020-01-01T08:25:37Z"),
        value: 63,
      },
      {
        customer_id: "tj@wheresmyemail.com",
        orderdate: DateTime.parse("2019-05-28T19:13:32Z"),
        value: 2,
      },
      {
        customer_id: "tj@wheresmyemail.com",
        orderdate: DateTime.parse("2020-11-23T22:56:53Z"),
        value: 187,
      },
      {
        customer_id: "tj@wheresmyemail.com",
        orderdate: DateTime.parse("2020-08-18T23:04:48Z"),
        value: 4,
      },
      {
        customer_id: "elise_smith@myemail.com",
        orderdate: DateTime.parse("2020-12-26T08:55:46Z"),
        value: 4,
      },
      {
        customer_id: "tj@wheresmyemail.com",
        orderdate: DateTime.parse("2021-02-28T07:49:32Z"),
        value: 1024,
      },
      {
        customer_id: "elise_smith@myemail.com",
        orderdate: DateTime.parse("2020-10-03T13:49:44Z"),
        value: 102,
      },
    ]
  )
  # end-insert-orders

  pipeline = [
    # start-match
    {
      "$match": {
        orderdate: {
          "$gte": DateTime.parse("2020-01-01T00:00:00Z"),
          "$lt": DateTime.parse("2021-01-01T00:00:00Z"),
        },
      },
    },
    # end-match
    # start-sort1
    {
      "$sort": {
        orderdate: 1,
      },
    },
    # end-sort1
    # start-group
    {
      "$group": {
        _id: "$customer_id",
        first_purchase_date: { "$first": "$orderdate" },
        total_value: { "$sum": "$value" },
        total_orders: { "$sum": 1 },
        orders: { "$push": {
          orderdate: "$orderdate",
          value: "$value",
        } },
      },
    },
    # end-group
    # start-sort2
    {
      "$sort": {
        first_purchase_date: 1,
      },
    },
    # end-sort2
    # start-set
    {
      "$set": {
        customer_id: "$_id",
      },
    },
    # end-set
    # start-unset
    { "$unset": ["_id"] },
    # end-unset
  ]

  # start-run-agg
  aggregation_result = orders.aggregate(pipeline)
  # end-run-agg

  aggregation_result.each do |doc|
    puts doc
  end
end

File: content/manual/v8.1/source/includes/aggregation/aggregation-examples/group-and-total/full-files/GroupTotal.scala
package org.example

import org.mongodb.scala.MongoClient
import org.mongodb.scala.bson.Document
import org.mongodb.scala.model._

import java.text.SimpleDateFormat

object GroupTotal {

  def main(args: Array[String]): Unit = {

    val uri = "<connection string>"
    val mongoClient = MongoClient(uri)
    Thread.sleep(1000)

    val aggDB = mongoClient.getDatabase("agg_tutorials_db")

    // start-insert-orders
    val orders = aggDB.getCollection("orders")

    orders.deleteMany(Filters.empty()).subscribe(
      _ => {},
      e => println("Error: " + e.getMessage),
    )

    val dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss")

    orders.insertMany(Seq(
      Document("customer_id" -> "elise_smith@myemail.com",
        "orderdate" -> dateFormat.parse("2020-05-30T08:35:52"),
        "value" -> 231),
      Document("customer_id" -> "elise_smith@myemail.com",
        "orderdate" -> dateFormat.parse("2020-01-13T09:32:07"),
        "value" -> 99),
      Document("customer_id" -> "oranieri@warmmail.com",
        "orderdate" -> dateFormat.parse("2020-01-01T08:25:37"),
        "value" -> 63),
      Document("customer_id" -> "tj@wheresmyemail.com",
        "orderdate" -> dateFormat.parse("2019-05-28T19:13:32"),
        "value" -> 2),
      Document("customer_id" -> "tj@wheresmyemail.com",
        "orderdate" -> dateFormat.parse("2020-11-23T22:56:53"),
        "value" -> 187),
      Document("customer_id" -> "tj@wheresmyemail.com",
        "orderdate" -> dateFormat.parse("2020-08-18T23:04:48"),
        "value" -> 4),
      Document("customer_id" -> "elise_smith@myemail.com",
        "orderdate" -> dateFormat.parse("2020-12-26T08:55:46"),
        "value" -> 4),
      Document("customer_id" -> "tj@wheresmyemail.com",
        "orderdate" -> dateFormat.parse("2021-02-28T07:49:32"),
        "value" -> 1024),
      Document("customer_id" -> "elise_smith@myemail.com",
        "orderdate" -> dateFormat.parse("2020-10-03T13:49:44"),
        "value" -> 102)
    )).subscribe(
      _ => {},
      e => println("Error: " + e.getMessage),
    )
    // end-insert-orders

    Thread.sleep(1000)

    val pipeline = Seq(
      // start-match
      Aggregates.filter(Filters.and(
        Filters.gte("orderdate", dateFormat.parse("2020-01-01T00:00:00")),
        Filters.lt("orderdate", dateFormat.parse("2021-01-01T00:00:00"))
      )),
      // end-match
      // start-sort1
      Aggregates.sort(Sorts.ascending("orderdate")),
      // end-sort1
      // start-group
      Aggregates.group(
        "$customer_id",
        Accumulators.first("first_purchase_date", "$orderdate"),
        Accumulators.sum("total_value", "$value"),
        Accumulators.sum("total_orders", 1),
        Accumulators.push("orders", Document("orderdate" -> "$orderdate", "value" -> "$value"))
      ),
      // end-group
      // start-sort2
      Aggregates.sort(Sorts.ascending("first_purchase_date")),
      // end-sort2
      // start-set
      Aggregates.set(Field("customer_id", "$_id")),
      // end-set
      // start-unset
      Aggregates.unset("_id")
      // end-unset
    )

    // start-run-agg
    orders.aggregate(pipeline)
      .subscribe((doc: Document) => println(doc.toJson()),
        (e: Throwable) => println(s"Error: $e"))
    // end-run-agg

    Thread.sleep(1000)
    mongoClient.close()
  }

}
